### PR TITLE
feat: initial polygon parts ingestion (MAPCO-4710)

### PR DESCRIPTION
**Pull request by vitaligi**
implement OpenAPI polygon-parts creation endpoint and functionality (first ingestion)

currently does not implements summary calculation

| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

### CODE REVIEW

**Review by CL-SHLOMIKONCHA — CHANGES_REQUESTED** [review]

File: src/polygonParts/models/interfaces.ts
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/interfaces.ts:1]: added on export interfaces

File: src/polygonParts/models/interfaces.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import type { IPolygonPart, PolygonPartsPayload } from '@map-colonies/mc-model-types';
+
+type PartData = Readonly<Omit<Pick<PolygonPartsPayload, 'partsData'>['partsData'][number], 'countries' | 'cities' | 'sensors'>> &
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/interfaces.ts:3]: used `PolygonPart` and renamed some interfaces

File: openapi3.yaml [hunk 2]
```diff
@@ -1,87 +1,333 @@
...
-    url: https://opensource.org/licenses/MIT
+tags:
+  - name: polygon-parts
+    description: CRUD operations to manage polygon-parts
+  - name: summary
+    description: Summary of polygon-parts data
 
 paths:
-  /anotherResource:
+  /summary/{catalogId}:
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply openapi3.yaml:13]: removed

File: openapi3.yaml [hunk 3]
```diff
@@ -1,87 +1,333 @@
...
-        - isAlive
+        - type
+      externalDocs:
+        url: http://geojson.org/geojson-spec.html#geometry-objects
+      properties:
+        type:
+          type: string
+          enum:
+            - Polygon
+            - MultiPolygon
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply openapi3.yaml:174]: removed

File: openapi3.yaml [hunk 4]
```diff
@@ -1,87 +1,333 @@
...
+    Point2D:
+      type: array
+      maxItems: 2
+      minItems: 2
+      items:
+        type: number
+    Summarized:
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply openapi3.yaml:181]: removed

File: openapi3.yaml [hunk 5]
```diff
@@ -1,87 +1,333 @@
...
+          type: number
+          format: double
+        sensors:
+          type: array
+          items:
+            type: string
+        footprint:
+          anyOf:
+            - $ref: '#/components/schemas/Polygon'
+            - $ref: '#/components/schemas/MultiPolygon'
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply openapi3.yaml:217]: removed

File: openapi3.yaml [hunk 6]
```diff
@@ -1,87 +1,333 @@
...
+      allOf:
+        - $ref: '#/components/schemas/GeometryBase'
+        - properties:
+            coordinates:
+              type: array
+              items:
+                type: array
+                items:
+                  $ref: '#/components/schemas/Point2D'
+    MultiPolygon:
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply openapi3.yaml:146]: removed

File: src/common/connectionManager.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+@singleton()
+export class ConnectionManager {
+  private appDataSource?: DataSource;
+
+  public constructor(@inject(SERVICES.LOGGER) private readonly logger: Logger, @inject(SERVICES.CONFIG) private readonly config: IConfig) {}
+
+  public static createConnectionOptions(dbConfig: DbConfig): DataSourceOptions {
+    const { enableSslAuth, sslPaths, ...connectionOptions } = dbConfig;
+    if (enableSslAuth && 'password' in connectionOptions && 'ssl' in connectionOptions) {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/common/connectionManager.ts:19]: eslint requires boolean expression

[the remark this answers is not in the document]
Comment (vitaligi) [reply src/common/connectionManager.ts:19]: i think the condition itself is incorrect and it should not check for these preoperties in `connectionOptions`, and it should only check `enableSslAuth`

File: src/common/connectionManager.ts [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    return connectionOptions;
+  }
+
+  public async init(): Promise<void> {
+    const connectionConfig = this.config.get<DbConfig>('db');
+    this.logger.info(
+      `connecting to database ${connectionConfig.database as string} ${connectionConfig.host !== undefined ? `on ${connectionConfig.host}` : ''}`
+    );
+    try {
+      if (this.appDataSource === undefined) {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/common/connectionManager.ts:32]: applied

File: config/custom-environment-variables.json [hunk 9]
```diff
@@ -38,5 +38,27 @@
...
+    "enableSslAuth": {
+      "__name": "DB_ENABLE_SSL_AUTH",
+      "__format": "boolean"
+    },
+    "sslPaths": {
+      "ca": "DB_CA_PATH",
+      "key": "DB_KEY_PATH",
+      "cert": "DB_CERT_PATH"
+    },
+    "database": "DB_NAME",
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply config/custom-environment-variables.json:63]: it is the same convention as in `raster-catalog-manager` and other repos

File: dataSource.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  return `${typeof tableOrName === 'string' ? tableOrName : tableOrName.name}_${columnNames.join('_')}${where !== undefined ? '_partial' : ''}_idx`;
+};
+customNamingStrategy.uniqueConstraintName = (tableOrName: Table | string, columnNames: string[]): string => {
+  return `${typeof tableOrName === 'string' ? tableOrName : tableOrName.name}_${columnNames.join('_')}`;
+};
+
+export const appDataSource = new DataSource({
+  ...ConnectionManager.createConnectionOptions(connectionOptions),
+  entities: ['src/**/DAL/*.ts'],
+  logging: true,
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply dataSource.ts:19]: changed the order of `DataSourceOptions` construction so now the "hardcoded" values are the default ones and the ones from pulled by `config` lib override them

File: dataSource.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  synchronize: false,
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply dataSource.ts:20]: see other comment

File: src/polygonParts/controllers/polygonPartsController.ts [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+type CreatePolygonPartsHandler = RequestHandler<undefined, string, PolygonPartsPayload>;
+
+const HTTP_STATUS_CREATED_TEXT = httpStatus.getStatusText(httpStatus.CREATED);
+
+@injectable()
+export class PolygonPartsController {
+  private readonly createdPolygonPartsCounter: BoundCounter;
+
+  public constructor(
+    @inject(SERVICES.LOGGER) private readonly logger: Logger,
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/controllers/polygonPartsController.ts:19]: removed

File: src/polygonParts/controllers/polygonPartsController.ts [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    @inject(SERVICES.METER) private readonly meter: Meter,
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/controllers/polygonPartsController.ts:20]: it was misused... now it is used

File: src/polygonParts/DAL/polygonPart.ts [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+@Entity({ name: 'polygon_parts', schema: 'polygon_parts', database: 'postgres' })
+@Unique('polygon_parts_insertion_order_uq', ['insertionOrder'])
+export class PolygonPart extends Common implements PolygonPartRecord {
+  @PrimaryGeneratedColumn('uuid', {
+    name: 'id',
+    primaryKeyConstraintName: 'polygon_parts_pkey',
+  })
+  public readonly id!: string;
+
+  @Column({ name: 'part_id', type: 'uuid' })
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/DAL/polygonPart.ts:14]: further DB behavior will be decided upon in the following PRs

File: .prettierignore [hunk 15]
```diff
@@ -1,3 +1,5 @@
+coverage
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply .prettierignore:1]: removed since it's a duplicate

File: src/common/errors.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import { NotFoundError, InternalServerError } from '@map-colonies/error-types';
+
+export class EntityNotFound extends NotFoundError {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/common/errors.ts:3]: it's unused
removing

File: src/polygonParts/models/polygonPartsManager.ts [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    this.logger.info(`creating polygon parts for catalog record: ${catalogId}`);
+
+    await this.connectionManager.getDataSource().transaction(async (entityManager) => {
+      const context: Context = {
+        catalogId,
+        entityManager,
+        resourceId,
+      };
+
+      await this.createSchema(context);
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:47]: renamed to `createTables` updated stored procedure name accordingly

File: src/polygonParts/models/polygonPartsManager.ts [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+    @inject(SERVICES.LOGGER) private readonly logger: Logger,
+    @inject(SERVICES.CONFIG) private readonly config: IConfig,
+    private readonly connectionManager: ConnectionManager
+  ) {
+    this.applicationConfig = config.get<ApplicationConfig>('application');
+  }
+
+  public async createPolygonParts(polygonPartsPayload: PolygonPartsPayload): Promise<void> {
+    const { catalogId } = polygonPartsPayload;
+    const resourceId = catalogId.replaceAll('-', '_');
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:36]: extracted to a function `getResourceId`

File: src/polygonParts/models/polygonPartsManager.ts
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:1]: using `InternalServerError` with a message

File: src/polygonParts/models/polygonPartsManager.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:36]: changed to `getEntityName`
naming should be tested with regex to comply with postgresql valid names

File: src/polygonParts/models/polygonPartsManager.ts [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+import { ApplicationConfig, IConfig } from '../../common/interfaces';
+import type { PolygonPartsIngestionPayload } from './interfaces';
+
+interface Context {
+  catalogId: PolygonPartsPayload['catalogId'];
+  entityManager: EntityManager;
+  resourceId: PolygonPartsPayload['catalogId'];
+}
+
+interface ErrorContext {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:17]: updated `js-logger` and using child logger

File: src/common/errors.ts [hunk 20]
```diff
@@ -0,0 +1,15 @@
...
+  public constructor(message: string) {
+    super(message);
+    Object.setPrototypeOf(this, EntityNotFound.prototype);
+  }
+}
+
+export class DBConnectionError extends InternalServerError {
+  public constructor() {
+    super('Internal Server Error');
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/common/errors.ts:5]: changed the behavior, now accepts a message and fallbacks to internal server error

File: src/polygonParts/models/polygonPartsManager.ts [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+        cities: partData.cities?.join(this.applicationConfig.arraySeparator),
+        description: partData.description,
+        geometry: partData.geometry,
+      };
+    });
+
+    try {
+      await entityManager.insert<PolygonPartsIngestionPayload>(`polygon_parts.${resourceId}_parts`, insertEntities);
+    } catch (error) {
+      const errorMessage = 'Could not insert polygon parts data';
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:124]: done

File: src/polygonParts/models/polygonPartsManager.ts [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+      throw new InternalServerError(this.enchanceErrorDetails({ error, errorMessage, id: catalogId }));
+    }
+  }
+
+  private async updatePolygonParts({ catalogId, entityManager, resourceId }: Context): Promise<void> {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:129]: done

File: src/polygonParts/models/polygonPartsManager.ts [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+    this.logger.debug(`updating polygon parts data for catalog record: ${catalogId}`);
+    try {
+      await entityManager.query(
+        `CALL "polygon_parts".update_polygon_parts('polygon_parts.${resourceId}_parts'::regclass, 'polygon_parts.${resourceId}'::regclass);`
+      );
+    } catch (error) {
+      const errorMessage = 'Could not update polygon parts data';
+      throw new InternalServerError(this.enchanceErrorDetails({ error, errorMessage, id: catalogId }));
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:137]: done

File: src/polygonParts/DAL/common.ts
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/DAL/common.ts:1]: why?

File: dataSource.ts [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+  migrations: ['src/db/migrations/*.ts'],
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply dataSource.ts]: see other comment

File: config/default.json [hunk 25]
```diff
@@ -24,5 +24,24 @@
...
+      "ca": "",
+      "key": "",
+      "cert": ""
+    },
+    "database": "postgres",
+    "schema": "polygon_parts",
+    "synchronize": false,
+    "logging": false,
+    "entities": ["**/DAL/*.js"],
+    "migrations": ["db/migrations/*.js", "db/migrations/*.ts"]
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply config/default.json:48]: now that this value is used for migration generation it should keep .ts

File: src/polygonParts/models/polygonPartsManager.ts [hunk 26]
```diff
@@ -0,0 +1,166 @@
...
+    const entityQualifiedName = this.getDatabaseObjectQualifiedName(entityName);
+
+    try {
+      if (insertEntities.length === 1) {
+        // QueryBuilder API is used since insert() of a single record uses the object keys as fields
+        // which is unsuitable since the keys have a mapping to column names
+        await entityManager
+          .createQueryBuilder()
+          .insert()
+          .into<IngestionProperties>(`${entityQualifiedName}_parts`, [
```
Comment (CL-SHLOMIKONCHA) [inline src/polygonParts/models/polygonPartsManager.ts:108]: Try to find any solution for this hard-coded columns

File: src/polygonParts/models/polygonPartsManager.ts [hunk 27]
```diff
@@ -4,68 +4,76 @@ import type { PolygonPartsPayload } from '@map-colonies/mc-model-types';
...
+  private readonly schema: DBSchema;
+
   public constructor(
     @inject(SERVICES.LOGGER) private readonly logger: Logger,
     @inject(SERVICES.CONFIG) private readonly config: IConfig,
     private readonly connectionManager: ConnectionManager
   ) {
-    this.applicationConfig = config.get<ApplicationConfig>('application');
+    this.applicationConfig = this.config.get<ApplicationConfig>('application');
+    this.schema = config.get<DbConfig['schema']>('db.schema') ?? DEFAULT_SCHEMA;
```
Comment (CL-SHLOMIKONCHA) [inline src/polygonParts/models/polygonPartsManager.ts:29]:
```suggestion
    this.schema = config.get<DBSchema>('db.schema') ?? DEFAULT_SCHEMA;
```

**Review by CL-SHLOMIKONCHA — CHANGES_REQUESTED** [review]

File: config/default.json [hunk 28]
```diff
@@ -24,5 +24,30 @@
         "options": null
       }
     }
+  },
+  "application": {
+    "arraySeparator": ",",
+    "createTablesStoredProcedure": "create_polygon_parts_tables",
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply config/default.json:30]: replaced

File: src/polygonParts/models/polygonPartsManager.ts [hunk 29]
```diff
@@ -0,0 +1,166 @@
...
+      const entityQualifiedName = this.getDatabaseObjectQualifiedName(entityName);
+      await entityManager.query(`CALL ${createPolygonPartsProcedure}('${entityQualifiedName}');`);
+    } catch (error) {
+      const errorMessage = 'Could not create polygon parts tables';
+      logger.error({ msg: errorMessage, error });
+      throw new InternalServerError(errorMessage);
+    }
+  }
+
+  private async insert(ingestionContext: IngestionContext): Promise<void> {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:69]: done

File: src/polygonParts/models/polygonPartsManager.ts
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:1]: changed

File: src/polygonParts/models/polygonPartsManager.ts (hunk 26 above)
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts]: it's now derived

**Review by CL-SHLOMIKONCHA — CHANGES_REQUESTED** [review]

File: src/polygonParts/models/polygonPartsManager.ts [hunk 30]
```diff
@@ -0,0 +1,220 @@
...
+import type { PolygonPartsPayload } from '@map-colonies/mc-model-types';
+import { inject, injectable } from 'tsyringe';
+import type { EntityManager } from 'typeorm';
+import { ConnectionManager } from '../../common/connectionManager';
+import { DEFAULT_SCHEMA, SERVICES } from '../../common/constants';
+import { ApplicationConfig, DbConfig, IConfig } from '../../common/interfaces';
+import { camelCaseToSnakeCase } from '../../common/utils';
+import type { IngestionProperties } from './interfaces';
+
+interface BaseIngestionContext {
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:12]: done

File: src/polygonParts/models/polygonPartsManager.ts [hunk 31]
```diff
@@ -0,0 +1,220 @@
...
+  private async verifyAvailableTableNames(ingestionContext: BaseIngestionContext): Promise<EntityNames> {
+    const { entityManager, logger, polygonPartsPayload } = ingestionContext;
+    const entityNames = this.getEntitiesNames(polygonPartsPayload);
+
+    logger.debug(`verifying polygon parts table names are available`);
+
+    await Promise.all(
+      Object.values<EntityName>({ ...entityNames }).map(async ({ databaseObjectQualifiedName, entityName }) => {
+        try {
+          const exists = await entityManager
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply src/polygonParts/models/polygonPartsManager.ts:79]: didn't found this capability :-(

File: openapi3.yaml [hunk 32]
```diff
@@ -93,38 +93,6 @@ components:
   schemas:
     PolygonPartsPayload:
```
[the remark this answers is not in the document]
Comment (vitaligi) [reply openapi3.yaml:94]: done

**Review by CL-SHLOMIKONCHA — APPROVED** [review]
Tests will be fix next PR

**Pull request merged**